Retain type sugar for extension declarations that name generic typealiases

DougGregor opened the pull request:
When extending a type via a generic typealias, where the type parameters of
the underlying nominal type line up precisely with those of the
generic typealias and its specialization of the underlying nominal
type (a so-called "pass-through" typealias in the new code), maintain
type sugar in the extension declaration.

This new type sugar enables inference of type requirements from the
generic typealias, which is both useful by itself (it lets the type
requirements on generic typealiases be meaningful for extensions like
they are elsewhere), and also addresses a source-compatability
regression where an extension of `CountableRange` will now infer the
requirement `Bound: Comparable`.

Fixes SR-6907 / rdar://problem/29066394.

DougGregor committed:
Retain type sugar for extension declarations that name generic typealiases.

When extending a type via a generic typealias, where the type parameters of
the underlying nominal type line up precisely with those of the
generic typealias and its specialization of the underlying nominal
type (a so-called "pass-through" typealias in the new code), maintain
type sugar in the extension declaration.

This new type sugar enables inference of type requirements from the
generic typealias, which is both useful by itself (it lets the type
requirements on generic typealiases be meaningful for extensions like
they are elsewhere), and also addresses a source-compatability
regression where an extension of `CountableRange` will now infer the
requirement `Bound: Comparable`.

Fixes SR-6907 / rdar://problem/29066394.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

slavapestov commented on test/Compatibility/CountableRange.swift:
```diff
@@ -0,0 +1,11 @@
+
+struct RequiresComparable<T: Comparable> { }
+
+extension CountableRange { // expected-warning{{'CountableRange' is deprecated: renamed to 'Range'}}
```
Can you test 'extension DictionaryIndex' too?

DougGregor replied:
Sure!

DougGregor committed:
DeclContext::getAsTypeOrTypeExtensionContext() generic typealiases.

This function was unable to "look through" an unbound generic typealias to
find the underlying nominal type that was getting extended, which can occur
if we end up performing a name lookup into this context before the extension
itself has been fully-checked.

Also, test that extending DictionaryIndex works.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor committed:
[Type checker] When we sugar an extension type, infer requirements.

The optimization that avoided forming new generic signatures for extensions
of nominal types was kicking in too eagerly, preventing us from inferring
requirements from the typealias.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please test source compatibility

DougGregor committed:
Update one last test for extensions of generic typealiases.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
Linux failure is weird:

```
15:17:41 CMake Error at CMakeLists.txt:90 (message):
15:17:41   LLDB test compilers not specified.  Tests will not run
15:17:41 
15:17:41 
15:17:41 -- Configuring incomplete, errors occurred!
```

lattner commented on lib/AST/DeclContext.cpp:
```diff
@@ -47,15 +47,33 @@ ASTContext &DeclContext::getASTContext() const {
+  auto type = ext->getExtendedType();
+  if (!type) return nullptr;
+
+  do {
```
Trivial style point, but a while(true) is arguably better than do/while(true) because you can know that it is an infinite loop reading the code top to bottom.

DougGregor replied:
Yeah, sure.

lattner commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -8222,46 +8222,139 @@ void TypeChecker::validateAccessControl(ValueDecl *D) {
+  // Check that the type parameters are the same the whole way through.
+  auto nominalGenericParams = nominalSig->getGenericParams();
+  auto typealiasGenericParams = typealiasSig->getGenericParams();
+  if (nominalGenericParams.size() != typealiasGenericParams.size())
```
Checking the sizes match is redundant with the subsequent std::equal call.

DougGregor replied:
Unfortunately, it isn't, because `std::equal` is terrible API that only takes the "first" iterator to the second sequence, assuming that one has already performed the size check ahead of time!

lattner replied:
Oh man.  🙄🙄

jrose-apple replied:
Well, there's the variant that takes the second end-iterator too…

lattner commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -8222,46 +8222,139 @@ void TypeChecker::validateAccessControl(ValueDecl *D) {
+  // a passthrough typealias.
+  auto innermostGenericParams = typealiasSig->getInnermostGenericParams();
+  auto boundArgs = boundGenericType->getGenericArgs();
+  if (boundArgs.size() != innermostGenericParams.size())
```
likewise.

lattner commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -8222,46 +8222,139 @@ void TypeChecker::validateAccessControl(ValueDecl *D) {
+  auto nominal = dyn_cast<NominalTypeDecl>(genericDecl);
+  auto typealias = dyn_cast<TypeAliasDecl>(genericDecl);
+  if (!nominal) {
+    Type underlyingType = typealias->getUnderlyingTypeLoc().getType();
```
Are you sure that genericDecl is either a typealias or a nominaltype decl?  If it is possible for it to be neither of those, then this code will crash with a null dereference.

lattner replied:
Ok, sounds good.  please add an assertion :)

lattner reviewed (commented):
LGTM, but I don't have this stuff paged into my head anymore.

DougGregor commented:
Okay, this one is actually good to go. The GRDB source-compatibility failure can't be addressed in a meaningful way. @lattner , I have another commit coming to fix the `while(true)` issue, but it's in a follow-on PR.

groue commented:
> The GRDB source-compatibility failure can't be addressed in a meaningful way.

I feel sorry (GRDB author here). Is there anything I can do to help?

DougGregor commented:
@groue actually, my comment was incorrect and I was confused! GRDB has provided a fantastic example here, and I'd missed that I needed to update the standard library to take advance of the compiler work in this pull request. The standard library change at https://github.com/apple/swift/pull/15513 gets GRDB passing again!

groue commented:
That's a relief, thank you Doug! Because the work that had to be done to make it pass was everything but trivial :-)

DougGregor commented:
You gave me an excuse to refactor something that I've felt is wrong/broken for ~2 years, so thanks!
